packages/react-components/src/localization/LocalizationProvider.tsx

```diff
@@ -76,10 +76,74 @@ export type LocalizationProviderProps = {
  * @remarks Components will be provided localized strings in English (US) by default if this
  * provider is not used
  */
-export const LocalizationProvider = (props: LocalizationProviderProps): JSX.Element => {
+const InternalLocalizationProvider = (props: InternalLocalizationProviderProps): JSX.Element => {
   const { locale, children } = props;
   return <LocaleContext.Provider value={locale}>{children}</LocaleContext.Provider>;
 };
 
 /** React hook to access locale */
 export const useLocale = (): ComponentLocale => useContext(LocaleContext);
+
+interface SetLocaleContext {
+  setLocale: (locale?: string) => Promise<void>;
+}
+
+const SetLocaleContext = createContext<SetLocaleContext>({
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars, @typescript-eslint/no-empty-function
+  setLocale: async (locale?: string) => {}
+});
+
+type AsyncLocalizationProviderProps = {
+  defaultLocale?: string;
+  localeLoader: (locale?: string) => Promise<ComponentLocale>;
+  children: React.ReactNode;
+};
+
+const AsyncLocalizationProvider = (props: AsyncLocalizationProviderProps): JSX.Element => {
+  const { children, defaultLocale, localeLoader } = props;
+  const [componentLocale, setComponentLocale] = useState<ComponentLocale>(COMPONENT_LOCALE_EN_US);
+
+  useEffect(() => {
+    const intializeLocale = async (): Promise<void> => {
+      const componentLocale = await localeLoader(defaultLocale);
+      setComponentLocale(componentLocale);
+    };
+    intializeLocale();
+  }, [defaultLocale, localeLoader]);
+
+  const state = useMemo<SetLocaleContext>(
+    () => ({
+      setLocale: async (locale?: string): Promise<void> => {
+        setComponentLocale(await localeLoader(locale));
+      }
+    }),
+    [localeLoader]
+  );
+
+  return (
+    <SetLocaleContext.Provider value={state}>
+      <InternalLocalizationProvider locale={componentLocale}>{children}</InternalLocalizationProvider>
+    </SetLocaleContext.Provider>
+  );
+};
+
+/**
+ * React hook for programmatically accessing the setLocale.
+ */
+export const useSetLocale = (): SetLocaleContext => useContext(SetLocaleContext);
+
+export const LocalizationProvider = (props: {
+  defaultLocale?: string;
+  locale: ComponentLocale | ((locale?: string) => Promise<ComponentLocale>);
+  children: React.ReactNode;
+}): JSX.Element => {
+  const { children, defaultLocale, locale } = props;
+  if (typeof locale === 'function') {
```

Comment:
I'd recommend against using typecheck to make logical decisions like this.

Since the synchronous and asynchronous locales work in very different ways, I'd argue it is better to keep the two separate in the API.
So, add two providers to the API: `AsyncLocalizationProvider` and `LocalizationProvider`. Internally, `LocalizationProvider` can make use of `AsyncLocalizationProvider` in the implementation like you do here.